### PR TITLE
Apiv3-7 home activities

### PR DESCRIPTION
**Pull request by sirambd**
Depends on #1072

### CODE REVIEW

File: app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt [hunk 1]
```diff
@@ -18,78 +18,121 @@
...
-        lastActivityJob.cancel()
-        lastActivityJob = Job()
+        forceDownload: Boolean = false,
+        isFirstPage: Boolean = false,
+        cursor: String? = null,
+    ) {
+        lastActivityJob?.cancel()
+        lastActivityJob = viewModelScope.launch(Dispatchers.IO) {
+            val lastActivityResult = getLastActivities(driveId, forceDownload, isFirstPage, cursor)
+            if (lastActivityJob?.isActive == true) lastActivitiesResult.postValue(lastActivityResult)
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt:73]: What is the difference with ensureActive() here ?

File: app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt [hunk 2]
```diff
@@ -18,78 +18,121 @@
...
+    ): LastActivityResult? {
 
         val ignoreDownload =
-            lastActivityPage == 1 && lastActivitiesTime != 0L && (Date().time - lastActivitiesTime) < DOWNLOAD_INTERVAL && !forceDownload
-        return liveData(Dispatchers.IO + lastActivityJob) {
-            if (ignoreDownload) {
-                lastActivityPage = lastActivityLastPage
-                emit(ApiResponse(SUCCESS, lastActivities, page = 1) to lastMergedActivities)
-                return@liveData
+            isFirstPage && lastActivitiesTime != 0L && (Date().time - lastActivitiesTime) < DOWNLOAD_INTERVAL && !forceDownload
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt:85]: I find it easier to with each && aligned to a new line

File: app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt [hunk 3]
```diff
@@ -18,78 +18,121 @@
...
+
+        if (ignoreDownload) {
+            return LastActivityResult(
+                mergedActivities = lastMergedActivities,
+                isComplete = false,
+                isFirstPage = true,
+            )
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt:92]: oneline this

File: app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt [hunk 4]
```diff
@@ -18,78 +18,121 @@
...
+        }
+
+        lastActivitiesTime = Date().time
+
+        val apiResponse = ApiRepository.getLastActivities(driveId, cursor)
+        lastActivityJob?.ensureActive()
+        val data = apiResponse.data
+
+        return if (apiResponse.isSuccess() || (isFirstPage && data != null)) {
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt:101]: Can we have an apiResponse that is error but still having data ? it seems a bit weird

File: app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt [hunk 5]
```diff
@@ -18,78 +18,121 @@
...
+        } else if (isFirstPage) {
+            val localActivities = FileController.getActivities()
+            val mergeAndCleanActivities = mergeAndCleanActivities(localActivities)
+            lastActivities = localActivities
+            lastMergedActivities = mergeAndCleanActivities
+            LastActivityResult(
+                mergedActivities = mergeAndCleanActivities,
+                isComplete = true,
+                isFirstPage = true,
+            )
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt:133]: oneline this too

File: app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt [hunk 6]
```diff
@@ -128,10 +171,11 @@ class HomeViewModel : ViewModel() {
                 previousActivity.fileId == currentActivity.fileId
     }
 
-    override fun onCleared() {
-        lastActivityJob.cancel()
-        super.onCleared()
-    }
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt:134]: Why did this disapear ?

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt (hunk 6 above)
Comment (sirambd) [reply app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt:134]: It's useless

File: app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt [hunk 7]
```diff
@@ -18,78 +18,121 @@
...
+                isFirstPage = true,
+            )
+        }
+
+        lastActivitiesTime = Date().time
+
+        val apiResponse = ApiRepository.getLastActivities(driveId, cursor)
+        lastActivityJob?.ensureActive()
+        val data = apiResponse.data
+
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt:100]: That return seems to be too complicated. Maybe we can simplify that part.

**Review by tevincent — CHANGES_REQUESTED** [review]
It would be great to improve that huge return. It's not readable.

File: app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt (hunk 1 above)
Comment (sirambd) [reply app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt:73]: `ensureActive` throw a CancellationException

Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt:71]: Edit: In this case, both have the same effect, but as the code is small and easy too read, isActive() was good here
(ensureActive() would need an extra ligne)

**Review by tevincent — APPROVED** [review]

**Review by FabianDevel — APPROVED** [review]

**Review by KevinBoulongne — APPROVED** [review]

**Pull request merged**